### ISSUE TITLE
[servant-client-0.5] generating composable client functions

### ISSUE THREAD

**Issue by echatav** [haskell-servant/servant#428]
Generating composable client functions with `servant-client-0.5` is difficult because of the `Manager` parameter, with the current workaround being the use of `unsafePerformIO`. This is mentioned in the [tutorial](http://haskell-servant.readthedocs.org/en/stable/tutorial/Client.html) and recently on [reddit](https://www.reddit.com/r/haskell/comments/4bw7ju/haskell_bindings_for_the_mondo_api_using_servant05/). I propose using a type family to distribute the `Manager` over alternatives, producing composable clients in the `ReaderT Manager (ExceptT ServantError IO)` monad. Here's a sketch. Let me know what you think.

```haskell
{-# LANGUAGE DataKinds #-}
{-# LANGUAGE FlexibleContexts #-}
{-# LANGUAGE FlexibleInstances #-}
{-# LANGUAGE MultiParamTypeClasses #-}
{-# LANGUAGE TypeFamilies #-}
{-# LANGUAGE TypeOperators #-}

module Distribute where

import Control.Monad.Except
import Control.Monad.Reader
import Data.Proxy
import Network.HTTP.Client (Manager)
import Servant.API
import Servant.Client

type family (-<>) a x where
  a -<> (b :<|> c) = (a -<> b) :<|> (a -<> c)
  a -<> (b -> c) = b -> (a -<> c)
  a -<> m b = ReaderT a m b

class Distribute a x where
  distribute :: (a -> x) -> (a -<> x)

instance {-# OVERLAPPING #-} (Distribute a b , Distribute a c) => Distribute a (b :<|> c) where
  distribute abc = distribute (altb . abc) :<|> distribute (altc . abc)
    where
      altb (b :<|> _) = b
      altc (_ :<|> c) = c

instance {-# OVERLAPPING #-} Distribute a c => Distribute a (b -> c) where distribute abc = distribute . flip abc

instance {-# OVERLAPPING #-} (a -<> m b) ~ (ReaderT a m b) => Distribute a (m b) where distribute = ReaderT

clients :: (HasClient layout, Distribute Manager (Client layout)) => Proxy layout -> BaseUrl -> (Manager -<> Client layout)
clients proxy baseurl = distribute (client proxy baseurl)

type API1 = "position" :> Capture "x" Int :> Capture "y" Int :> Get '[JSON] (Int,Int)
type API2 = "hello" :> QueryParam "name" String :> Get '[JSON] String
type API3 = "goodbye" :> QueryParam "name" String :> Get '[JSON] String
type API = API1 :<|> API2 :<|> API3

api :: Proxy API
api = Proxy

baseUrl :: BaseUrl
baseUrl = BaseUrl Http "localhost" 8081 ""

positionClient :: Int -> Int -> ReaderT Manager (ExceptT ServantError IO) (Int, Int)
helloClient :: Maybe String -> ReaderT Manager (ExceptT ServantError IO) String
goodbyeClient :: Maybe String -> ReaderT Manager (ExceptT ServantError IO) String
positionClient :<|> helloClient :<|> goodbyeClient = clients api baseUrl
```

**Comment by alpmestan** [comment]
Isn't it possible to write this "transformation" with `enter`?

**Comment by echatav** [comment]
`(->) a` and `(-<>) a` are indeed naturally isomorphic. We could write `distribute' :: ((->) a) :~> ((-<>) a)` with `distribute' = Nat distribute`, and then we'd have `enter distribute' = distribute`. I'm not sure how that'd be useful. What I'm trying to solve here is not how to use another monad. The `ReaderT` bit is a suggestion from @aaronlevin from that reddit thread and helps with composability and sharing a single `Manager` but is not the main thing. The main thing is generating clients without plugging in an `unsafePerformIO (newManager defaultManagerSetting)`. From what I understand `enter` is used for changes like `m a :<|> m b` to `n a :<|> n b` whereas I'm trying to get a change like `m (a :<|> b)` -> `m a :<|> m b`.

**Comment by echatav** [comment]
The `servant-0.5` [announcement](http://haskell-servant.github.io/posts/2016-03-19-servant-0.5-release.html#manager-and-baseurl-as-an-argument) also describes the problem.

**Comment by mbg** [comment]
I am the author of the reddit post that was linked to by @echatav. I think he describes the problem perfectly and his code does solve it. I am just wondering why the `Manager` parameter is part of `client` in the first place since I am assuming that the manager isn't actually used in the generation of the client functions. So in other words, right now `client` is generating a bunch of functions which are specific to some HTTP client. If I wanted to use two different HTTP clients, I would have to generate two sets of client functions. Wouldn't `Manager -> ExceptT ServantError IO a` or `ReaderT Manager (ExceptT ServantError IO a` be a more sensible default for the client functions so that the `Manager` parameter could be removed from `client`?

**Comment by alpmestan** [comment]
Maybe it's time we just put together a list of requirements for the `client` machinery and then implement a solution to it? We've been tweaking it in different directions since it was born but we kept being short of a solution for other problems. I'm sure we can find a decent solution that allows people who write client libraries to existing APIs to have an easy life. We've got all the essential bits in place already, it's just about wrapping it all nicely.

**Comment by soenkehahn** [comment]
@echatav: Some time ago I had a discussion with some servant contributors about having general helper functions to work with API types and `:<|>`-structures (on the value level) in general. Your `-<>` looks like it's close to one of those helper functions. Do you think it could be generalized to not only distribute arguments into `ReaderT`s but generally transform leaves in `:<|>`-structures?
If yes, I think this should live in the `servant` package to be available everywhere. Would putting this into `servant` solve your use-case?

And personally I find `-<>` not a good name choice. I wouldn't pick an operator. Maybe something like `Distribute` for the type family and `CanDistribute` for the class?

**Comment by echatav** [comment]
There are some possibilities for generalization. For instance, the class can also have an `undistribute` function as it's an isomorphism. However, it may be difficult to anticipate how to structure a more general type family. For instance, I wrote two versions of it; one which only distributes `((->) a)` over `:<|>`s and the one above which also pushes the `a` to be the last parameter so that it can then be transformed into a `ReaderT`. The `ReaderT` version is in some sense less general, but is informed by a use case. It may be decomposed into two type families, a `Distribute f a` with `Distribute f (a :<|> b) ~ (f a) :<|> (Distribute f b)` and `ReaderTify a b`. With the more general `Distribute`, in addition to `CanDistribute ((->) a)` you could define `CanDistribute ((:>) a)` which would be useful for more compactly defining apis that have a lot of repetition. There's lots of interesting things to think about here, but maybe something like the code above should be added as a stop gap before any grand redesign of `servant-client` so that users can more easily develop client libraries for existing web apis.

**Comment by echatav** [comment]
I think that a general `Distribute` type family combines well with `Enter` to accomplish @soenkehahn's goal of transforming leaves in `:<|>`-structures.
```haskell
type family Distribute f a where
  Distribute f (a :<|> b) = (Distribute f a) :<|> (Distribute f b)
  Distribute f a = f a
class CanDistribute f where
  distribute :: f a -> Distribute f a
```
Then, given `t :: f :~> g`, you can `enter t . distribute :: CanDistribute f => f (a :<|> b :<|> c) -> (g a :<|> g b :<|> g c)`. I am a bit confused about what may happen when there's left parenthesized alternatives though.

**Comment by echatav** [comment]
@mbg's suggestion is much cleaner than my proposal. Change the signature of `clientWithRoute` to `Proxy layout -> Req -> BaseUrl -> Client layout`, removing `Manager` and instead putting it in the `Client` type, e.g. `type Client (Verb method status cts NoContent) = ReaderT Manager (ExceptT ServantError IO) NoContent`.

**Comment by soenkehahn** [comment]
Some versions ago we had the `BaseUrl` parameter being passed in like @mbg suggests for the `Manager`. So e.g. `Client (Get '[JSON] Something) ~ (BaseUrl -> ExceptT ServantError IO Something)`. I liked that much better, since sometimes the `BaseUrl` is not known at compile time and then you have the same problem as you have with `Manager` now.

If anyone wants to create a PR moving `Manager` into the `Client` type (or even `Manager` and `BaseUrl`) I'd support merging that. (And I would prefer using function types instead of `ReaderT`s.)

**Comment by mbg** [comment]
I could have a go at implementing this later in the week. 

I am not sure whether `ReaderT` or function types would be the better choice, though. I am guessing that most clients would use the same `BaseUrl` and `Manager` throughout the code, so the question is whether we pass on the responsibility of keeping track of those two values to people using `servant-client` or not. E.g. in my library, I'd end up wrapping it up in `ReaderT` anyway and I'd still have to wrap every function returned by `client` into a wrapper (which I don't mind doing), but there is a question here about whether we think most people will end up doing something similar (in which case we should probably go with `ReaderT` out of the box) or not (in which case function types would be preferable).

**Comment by soenkehahn** [comment]
Good point. I'm not very passionate about the `ReaderT` thing, so I'd leave that decision up to you (or whoever ends up implementing this).

Of course, ideally at some point we would still have a way to generically modify all the leaves in a `:<|>`-structure and then we hopefully can easily transform the `ReaderT` variant into the function variant and vice versa.

**Comment by soenkehahn** [comment]
@echatav: Since #432 is merged, can this be closed?

**Comment by echatav** [comment]
Solved in #432
